src/flamenco/gossip/fd_gossip.c

```diff
@@ -1,215 +1,312 @@
-    fd_crds_value_t * candidate = fd_crds_mask_iter_value( it );
-
-    /* TODO: Add jitter here? */
-    if( FD_UNLIKELY( fd_crds_value_wallclock( candidate )>contact_info->wallclock ) ) continue;
-
-    ulong serialized_sz;
-    error = serialize_crds_value_into_packet( candidate, packet, 1232UL-packet_sz, &serialized_sz );
-    if( FD_LIKELY( !error ) ) {
-      packet_sz += serialized_sz;
-    } else {
-      /* CRDS value can't fit into the packet anymore, just ship what
-         we have now and start a new one. */
-      gossip->tx_fn( gossip->tx_ctx, packet, packet_sz );
-      packet_sz = 0UL;
-    }
+verify_signatures( fd_gossip_message_t const *  message,
+                   uchar const *                payload ) {
+  if( FD_UNLIKELY( !( !!message->has_signable_data || !!message->crds_cnt ) ) )
+    return FD_GOSSIP_RX_VERIFY_NO_SIGNABLE_DATA;
+  
+  fd_sha512_t sha[1];
+
+  int err = 0;
+  /* Optimize for CRDS composites (push/pull) that don't have an outer signable
+     data */
+  if( FD_UNLIKELY( message->has_signable_data ) ) {
+    /* TODO: Special case for prune */
+    err = fd_ed25519_verify( payload + message->signable_data_offset,
+                             message->signable_sz,
+                             message->signature,
+                             message->pubkey,
+                             sha );
   }
-
-  /* TODO: Send packet if there's anything leftover */
-
-  return 0;
-}
-
-static int
-rx_pull_response( fd_gossip_t *                     gossip,
-                  fd_gossip_pull_response_t const * pull_response,
-                  long                              now ) {
-  /* TODO: use epoch_duration and make timeouts ... ? */
-
-  for( ulong i=0UL; i<pull_response->values_len; i++ ) {
-    int upserts = fd_crds_upserts( gossip->crds, pull_response->values[ i ] );
-
-    if( FD_UNLIKELY( !upserts ) ) {
-      failed_inserts_append( gossip, pull_response->values[ i ] );
-      continue;
-    }
-
-    /* TODO: Is this jittered in Agave? */
-    long accept_after_nanos;
-    if( FD_UNLIKELY( !memcmp( pull_response->sender_pubkey, gossip->identity_pubkey, 32UL ) ) ) {
-      accept_after_nanos = 0L;
-    } else if( stake( pull_response->sender_pubkey ) ) {
-      accept_after_nanos = now-15L*1000L*1000L*1000L;
-    } else {
-      accept_after_nanos = now-432000L*1000L*1000L*1000L;
-    }
-
-    if( FD_LIKELY( accept_after_nanos<=fd_crds_value_wallclock( pull_response->values[ i ] ) ) ) {
-      fd_crds_insert( gossip->crds, pull_response->values[ i ], now );
-      fd_crds_update_record_timestamp( pull_response->sender_pubkey, now );
-    } else if( fd_crds_has_contact_info( pull_response->sender_pubkey ) ) {
-      fd_crds_insert( gossip->crds, pull_response->values[ i ], now );
-    } else {
-      failed_inserts_append( gossip, pull_response->values[ i ] );
-    }
+  if( FD_UNLIKELY( err != FD_ED25519_SUCCESS ) ) return err;
```

Comment:
kbowers style nit, :) please apply everywhere

```suggestion
  if( FD_UNLIKELY( err!=FD_ED25519_SUCCESS ) ) return err;
```